File: python/a_star.py
from typing import List, Optional

from board import Board
from position import Position
from heuristics import Heuristic
from path import Path


class AStarSearch:
    def __init__(self, board: Board, heuristic: Heuristic) -> None:
        self.board = board
        self.heuristic = heuristic

    def _init_attrs(self, start: Position, goal: Position) -> None:
        self._open_set = set([start])
        self._closed_set = set()
        self._came_from = {}
        self._g_score = {start: 0}
        self._f_score = {start: self.heuristic.get_distance(start, goal)}

    def find_path(
        self, start: Optional[Position] = None, goal: Optional[Position] = None
    ) -> Path:
        """Returns the path as a list of positions."""
        if start is None:
            start = self.board.get_start()
        if goal is None:
            goal = self.board.get_end()
        self._init_attrs(start, goal)

        while len(self._open_set) > 0:
            current = self._find_min_f_position()

            if current == goal:
                return self._reconstruct_path(goal)

            self._open_set.remove(current)
            self._closed_set.add(current)

            for neighbour in self.board.get_neighbours(current):
                if neighbour in self._closed_set:
                    continue
                if neighbour not in self._open_set:
                    self._open_set.add(neighbour)

                self._appoint_f_score(current, neighbour, goal)

    def _find_min_f_position(self) -> Position:
        min_f_pos = next(iter(self._open_set))
        min_f_score = self._f_score[min_f_pos]

        for pos in self._open_set:
            if self._f_score[pos] < min_f_score:
                min_f_score = self._f_score[pos]
                min_f_pos = pos
        return min_f_pos

    def _appoint_f_score(
        self, current: Position, neighbour: Position, goal: Position
    ) -> None:
        tentative_g_score = self._g_score[current] + self.board.get_cost(
            neighbour
        )
        self._came_from[neighbour] = current
        self._g_score[neighbour] = tentative_g_score
        self._f_score[
            neighbour
        ] = tentative_g_score + self.heuristic.get_distance(neighbour, goal)

    def _reconstruct_path(self, position: Position) -> Path:
        positions = [position]
        cost = self._f_score[position]
        while position in self._came_from:
            position = self._came_from[position]
            positions.insert(0, position)
        path = Path(positions, cost)
        return path

    def __str__(self) -> str:
        return f"A* search algorithm with {self.heuristic()}"
    

File: java/AStar.java
import java.util.Set;
import java.util.ArrayList;
import java.util.HashMap;
import java.util.HashSet;
import java.util.List;
import java.util.Map;

public class AStar {
    private Board board;
    private Heuristic heuristic;
    private Set<Position> openSet;
    private Set<Position> closedSet;
    private Map<Position, Position> cameFrom;
    private Map<Position, Integer> gScore;
    private Map<Position, Integer> fScore;

    public AStar(Board board, Heuristic heuristic) {
        this.board = board;
        this.heuristic = heuristic;
    }

    private void initAttrs(Position start, Position goal) {
        openSet.add(start);
        closedSet = new HashSet<Position>();
        cameFrom = new HashMap<Position, Position>();
        gScore = new HashMap<Position, Integer>();
        gScore.put(start, 0);
        fScore = new HashMap<Position, Integer>();
        fScore.put(start, this.heuristic.getDistance(start, goal));
    }

    public List<Position> findPath(Position start, Position goal) {
        initAttrs(start, goal);
        while (openSet.size() > 0) {
            Position current = findMinFPosition();

            if (current == goal) return reconstructPath(goal);
            openSet.remove(current);
            closedSet.add(current);

            for (Position n : board.getNeighbours(current)) {
                if (closedSet.contains(n)) continue;
                if (!openSet.contains(n)) openSet.add(n);
                appointFScore(current, n, goal);
            }
        }
        List<Position> emptyList = new ArrayList<Position>();
        return emptyList;
    }

    private Position findMinFPosition() {
        Position minFPos = openSet.toArray(new Position[openSet.size()])[0];
        int minFScore = fScore.get(minFPos);
        for (Position pos : openSet) {
            if (fScore.get(pos) < minFScore) {
                minFScore = fScore.get(pos);
                minFPos = pos;
            }
        }
        return minFPos;
    }

    private void appointFScore(Position current, Position neighbour, Position goal) {
        int tentativeGScore = gScore.get(current) + board.getCost(neighbour);
        cameFrom.put(neighbour, current);
        gScore.put(neighbour, tentativeGScore);
        fScore.put(neighbour, tentativeGScore + heuristic.getDistance(neighbour, goal));
    }

    private List<Position> reconstructPath(Position position) {
        List<Position> positions = new ArrayList<Position>();
        positions.add(position);
        // int cost = fScore.get(position);
        while (cameFrom.containsKey(position)) {
            position = cameFrom.get(position);
            positions.add(position);
        }
        return positions;
    }
}
